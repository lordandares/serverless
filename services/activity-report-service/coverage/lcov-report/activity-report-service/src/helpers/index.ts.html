<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for activity-report-service/src/helpers/index.ts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../index.html">All files</a> / <a href="index.html">activity-report-service/src/helpers</a> index.ts
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>34/34</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/0</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>18/18</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20</td><td class="line-coverage quiet"><span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import * as s3 from './s3'
&nbsp;
export { buildLocationSummaryPdf } from './build-location-summary-pdf'
export { buildShiftSummaryPdf } from './build-shift-summary-pdf'
export { buildEmail } from './build-email'
export { buildShiftEmail } from './build-shift-email'
export { getShiftDetails } from './get-shift-details'
export { getShiftReportLinks } from './get-shift-report-links'
export { getUserDetails } from './get-user-details'
export { concatPdfs } from './concat-pdfs'
export { downloadFiles } from './download-files'
export { fetchApplications } from './fetch-applications'
export { fetchLocations } from './fetch-locations'
export { getLocationReportLinks } from './get-location-report-links'
export { getShiftSummaryData } from './get-shift-summary-data'
export { getSummaryData } from './get-summary-data'
export { getFormS3Paths } from './get-references'
export { s3 }
export { sendEmail } from './send-email'
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Mon Jan 13 2020 21:52:44 GMT+0000 (Coordinated Universal Time)
</div>
</div>
<script src="../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../sorter.js"></script>
</body>
</html>
