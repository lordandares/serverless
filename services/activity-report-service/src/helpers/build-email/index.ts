import * as BPromise from 'bluebird'
import {
  filter,
  find,
  flow,
  get,
  getOr,
  isEmpty,
  reject,
  toUpper,
  trim,
} from 'lodash/fp'
import * as moment from 'moment-timezone'

interface Env {
  env: {
    STAGE: string
    SES_CONFIGURATION_SET_NAME: string
  }
}

declare var process: Env

import { Location } from '../../handlers/email-processor'
import { getLocationReportLinks } from '../get-location-report-links'
import { UserDetails } from '../get-user-details'

const getFormSubmissionOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(true, 'options.formSubmissions'),
)

const getNoActivityOption = flow(
  get('notifications.channels'),
  find({ name: 'daily-location-report' }),
  getOr(false, 'options.skipNoActivity'),
)

export interface BuildEmail extends UserDetails {
  locations: Location[]
}

export async function buildEmail(options: BuildEmail) {
  try {
    const {
      applicationName,
      email,
      firstName,
      locations,
      preferences,
      timezone,
    } = options

    const isProduction = process.env.STAGE === 'production'
    const stage = toUpper(process.env.STAGE)
    const template = isProduction
      ? 'daily-location-reports'
      : 'daily-location-reports-test'

    const ConfigurationSetName = process.env.SES_CONFIGURATION_SET_NAME

    const mStartDate = moment()
      .utc()
      .tz(timezone)
      .subtract(1, 'day')

    const dateText = mStartDate.format('dddd Do MMMM')
    const shortDateText = mStartDate.format('MMMM D')

    const withFormSubmissions = getFormSubmissionOption(preferences)
    const skipNoActivity = getNoActivityOption(preferences)

    const links = await BPromise.map(locations, location =>
      getLocationReportLinks({
        location,
        withFormSubmissions,
      }),
    )

    const activeLocations = filter('link')(links)
    const inactiveLocations = reject('link')(links)

    const hasNoLocations =
      isEmpty(activeLocations) && isEmpty(inactiveLocations)

    if (
      isEmpty(activeLocations) &&
      !isEmpty(inactiveLocations) && skipNoActivity
    ) {
      console.info('skipping activity report', {
        activeLocations: activeLocations.length,
        inactiveLocations: inactiveLocations.length,
        skipNoActivity,
      })
      return false
    }

    const name = trim(firstName)

    const data = JSON.stringify({
      activeLocations,
      applicationName,
      date: dateText,
      hasNoLocations,
      inactiveLocations,
      isProduction,
      name,
      sendInactive: !skipNoActivity,
      shortDate: shortDateText,
      stage,
    })

    return {
      ...(ConfigurationSetName && { ConfigurationSetName }),
      Destination: {
        CcAddresses: [],
        ToAddresses: [email],
      },
      ReplyToAddresses: ['notifications@lighthouse.io'],
      Source: 'notifications@lighthouse.io',
      Template: template,
      TemplateData: data,
    }
  } catch (error) {
    console.error('Error building email', { error, options })
    return false
  }
}
