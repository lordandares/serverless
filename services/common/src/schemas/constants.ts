// ---------------- //

export const GROUP_LOCATION = 'location'
export const GROUP_OCCURRENCE = 'occurrence'
export const GROUP_SCHEDULE = 'schedule'

export const groupTypes = [GROUP_LOCATION, GROUP_OCCURRENCE, GROUP_SCHEDULE]

// ---------------- //

export const STATUS_ACTIVE = 'active'
export const STATUS_PENDING = 'pending'
export const STATUS_RESOLVED = 'resolved'
export const STATUS_EXPIRED = 'expired'
export const STATUS_DELETED = 'deleted'

export const statuses = [
  STATUS_ACTIVE,
  STATUS_PENDING,
  STATUS_RESOLVED,
  STATUS_EXPIRED,
  STATUS_DELETED,
]

// ---------------- //

export const EVENT_VISIT = 'visit'

export const eventTypes = [EVENT_VISIT]
