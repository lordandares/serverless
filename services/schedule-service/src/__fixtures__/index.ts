import { Context } from 'aws-lambda'

export const areaDocument = {
  _id: 'location1',
  address: {
    city: 'Bethesda',
    country: 'US',
    postal: '20889',
    state: 'MD',
    street: '8901 Rockville Pike',
    street2: '',
  },
  altName: 'WRNMMC Facility',
  application: '56789671d4c6488b128c7f9c',
  areaSize: 536559.55514508,
  deleted: false,
  floorsRef: [0],
  groupType: 'location',
  keywords: ['hospital'],
  name: 'WRNMMC Facility',
  serviceHours: {
    hours: [
      {
        description: 'MON 00:00 - TUE 00:00',
        duration: 86400000,
        end: 86400000,
        start: 0,
        type: 'DEFAULT',
      },
    ],
    timezone: 'America/New_York',
  },
  tags: [],
  timezone: 'America/New_York',
}

export const occurrenceDocument = {
  applicationId: 'application1',
  createdAt: '2019-01-01T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    occurrenceInterval: [1, 2],
    scheduleName: 'Schedule 1',
    serviceInterval: [1, 2],
    timezone: 'Australia/Melbourne',
    type: 'visit',
  },
  endAt: '2019-02-01T23:59:59.999Z',
  groupType: 'occurrence',
  locationId: '5d163137b8b3b7000127edd1',
  location_endAt_occurrenceId: 'location1-2019-02-02T00:00:00.000Z-occurrence1',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-01-01T00:00-occurrence1',
  startAt: '2019-02-01T00:00:00.000Z',
  status: 'pending',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}

export const occurrenceTimerDocument = {
  pk: `timer-2019-01-01T00:00`,
  sk: `timer#application1-occurrence#2019-01-01T00:00-occurrence`,
  endAt: '2019-01-01T23:59:59.999Z',
  groupType: `occurrence-timer`,
  scheduleId: 'schedule1',
  targetArn: 'someArn',
}

export const scheduleDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
  data: {
    areas: ['area1'],
    enabled: true,
    included: {},
    locations: ['location1'],
    name: 'Schedule One',
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
    strategy: {
      options: {
        duration: {
          unit: 'minute',
          value: 15,
        },
        frequency: {
          unit: 'minute',
          value: 15,
        },
      },
      type: 'stopwatch',
    },
    type: 'visit',
  },
  endAt: '2019-10-1500:00:00.000Z',
  groupType: 'schedule',
  pk: 'application1-schedule',
  scheduleId: 'schedule1',
  sk: '2019-10-13T00:00:00.000Z-schedule1',
  startAt: '2019-10-1400:00:00.000Z',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: 'user1',
    label: 'Unknown User',
    type: 'user',
  },
}

export const scheduleLocationDocument = {
  applicationId: 'application1',
  createdAt: '2019-10-13T00:00:00.000Z',
  createdBy: {
    id: '123456789',
    label: 'label',
    type: 'system',
  },
  data: {
    name: 'Schedule One',
    serviceHours: {
      hours: [
        {
          description: 'MON 11:00 - TUE 00:00',
          end: 86400000,
          start: 39600000,
          type: 'DEFAULT',
        },
      ],
      timezone: 'Australia/Melbourne',
    },
  },
  groupType: 'location',
  pk: 'application1-location',
  schedules: ['schedule1'],
  sk: 'location1',
  updatedAt: '2019-10-13T01:00:00.000Z',
  updatedBy: {
    id: '123456789',
    label: 'label',
    type: 'system',
  },
}

export const schedulePayload = {
  areas: ['area1'],
  enabled: true,
  endAt: '2019-10-1500:00:00.000Z',
  locations: ['location1'],
  name: 'Schedule One',
  serviceHours: {
    hours: [
      {
        description: 'MON 11:00 - TUE 00:00',
        end: 86400000,
        start: 39600000,
        type: 'DEFAULT',
      },
    ],
    timezone: 'Australia/Melbourne',
  },
  startAt: '2019-10-1400:00:00.000Z',
  strategy: {
    options: {
      duration: {
        unit: 'minute',
        value: 15,
      },
      frequency: {
        unit: 'minute',
        value: 15,
      },
    },
    type: 'stopwatch',
  },
  type: 'visit',
}

export const rulePayload = {
  applicationId: 'application-id-1111',
  data: {
    occurrenceInterval: [1580761454998, 1580762054997],
  },
  hook: {
    endpoint: 'sns:endpoint:uri',
    type: 'sns',
  },
  locationId: 'location-id-11111',
  occurrenceId: 'occurrence1',
  pk: 'application1-occurrence',
  scheduleId: 'schedule1',
  sk: '2019-06-02T00:00:00.000Z-occurrence1',
  startAt: '2019-02-01T00:00:00.000Z',
  type: 'visit',
}

export const rulePatternDocument = {
  createdAt: '2019-06-02T00:00:00.000Z',
  createdBy: {
    id: 'created-by-id-11111',
    label: 'created-by-label-1111',
    type: 'created-by-type-11111',
  },
  groupType: 'rule-pattern',
  matches: {
    occurrence1: {
      pk: 'application1-occurrence',
      sk: '2019-06-01T00:00:00.000Z-occurrence1',
    },
  },
  pk: 'rule-pattern-application-id-11111-location-id-11111',
  sk: 'visit',
  updatedAt: '2019-06-02T00:00:00.000Z',
}

export const context: Context = {
  awsRequestId: 'aws-request-id',
  callbackWaitsForEmptyEventLoop: false,
  clientContext: undefined,
  done: jest.fn(),
  fail: jest.fn(),
  functionName: 'function-name',
  functionVersion: 'function-version',
  getRemainingTimeInMillis: jest.fn(),
  identity: undefined,
  invokedFunctionArn: 'invoked-function-arn',
  logGroupName: 'log-group-name',
  logStreamName: 'log-stream-name',
  memoryLimitInMB: 0,
  succeed: jest.fn(),
}
