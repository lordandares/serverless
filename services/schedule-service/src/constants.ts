export const MAX_OCCURRENCES = 2
export const MAX_STOPWATCH_OCCURRENCES = 1

export const OCCURRENCE_STATUS_ACTIVE = 'active'
export const OCCURRENCE_STATUS_PENDING = 'pending'
export const OCCURRENCE_STATUS_EXPIRED = 'expired'
export const OCCURRENCE_STATUS_RESOLVED = 'resolved'
