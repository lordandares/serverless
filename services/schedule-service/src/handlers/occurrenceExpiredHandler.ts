import { errors, schemas } from '@lighthouse/serverless-common'
import { Context, SNSEvent } from 'aws-lambda'
import { DynamoDB, SNS, StepFunctions } from 'aws-sdk'
import { DocumentClient as DC } from 'aws-sdk/clients/dynamodb'

import {
  OCCURRENCE_STATUS_ACTIVE,
  OCCURRENCE_STATUS_EXPIRED,
} from '../constants'
import parseSnsEvent from './lib/parseSnsEvent'

declare var process: {
  env: {
    CREATE_EXCEPTION_ARN: string
    ENSURE_OCCURRENCES_STEP_FUNCTION: string
    TABLE_SCHEDULES: string
  }
}

export async function occurrenceExpiredHandler(
  event: SNSEvent,
  context: Context,
): Promise<void> {
  try {
    const { body } = parseSnsEvent(event)

    if (!body) {
      throw new errors.ApplicationError({
        message: 'occurrenceExpiredHandler: missing body on event',
      })
    }

    const { pk: occurrencePk, sk: occurrenceSk } = body
    const { awsRequestId, invokedFunctionArn } = context

    const documentClient = new DynamoDB.DocumentClient()
    const sfs = new StepFunctions()
    const snsClient = new SNS()
    const tableName = process.env.TABLE_SCHEDULES

    const occurrenceGetParams: DC.GetItemInput = {
      Key: { pk: occurrencePk, sk: occurrenceSk },
      TableName: tableName,
    }

    const occurrenceGetResponse: DC.GetItemOutput = await documentClient
      .get(occurrenceGetParams)
      .promise()

    const { Item: occurrenceDocument } = occurrenceGetResponse

    if (!occurrenceDocument) {
      throw new errors.ResourceNotFoundError({
        id: JSON.stringify({ pk: occurrencePk, sk: occurrenceSk }),
        resource: 'occurrence',
      })
    }

    console.debug('OccurrenceExpiredHandler: setting occurrence to expired', {
      occurrence: JSON.stringify(occurrenceDocument),
    })

    const {
      occurrenceId,
      status,
    } = occurrenceDocument as schemas.ScheduleOccurrenceDocumentSchema

    if (status !== OCCURRENCE_STATUS_ACTIVE) {
      console.info(
        `OccurrenceExpiredHandler: occurrence is no longer active and has ${status} status so skipping`,
        {
          occurrenceId,
        },
      )

      return
    }

    const timestamp = new Date().toISOString()

    const updateParams: DC.UpdateItemInput = {
      AttributeUpdates: {
        expiredAt: {
          Action: 'PUT',
          Value: timestamp,
        },
        expiredBy: {
          Action: 'PUT',
          Value: invokedFunctionArn,
        },
        status: {
          Action: 'PUT',
          Value: OCCURRENCE_STATUS_EXPIRED,
        },
        updatedAt: {
          Action: 'PUT',
          Value: timestamp,
        },
        updatedBy: {
          Action: 'PUT',
          Value: {
            id: awsRequestId,
            label: invokedFunctionArn,
            type: 'system',
          },
        },
      },
      Key: {
        pk: occurrencePk,
        sk: occurrenceSk,
      },
      ReturnValues: 'ALL_NEW',
      TableName: tableName,
    }

    console.info('OccurrenceExpiredHandler: updating occurrence status')

    const occurrenceUpdateResponse: DC.UpdateItemOutput = await documentClient
      .update(updateParams)
      .promise()

    console.info('OccurrenceExpiredHandler: set occurrence status to expired')

    const { Attributes: updatedOccurrenceDocument } = occurrenceUpdateResponse

    console.info(
      `OccurrenceExpiredHandler: sending message to ARN: ${
        process.env.CREATE_EXCEPTION_ARN
      }`,
    )

    const snsResponse = await snsClient
      .publish({
        Message: JSON.stringify(updatedOccurrenceDocument),
        TopicArn: process.env.CREATE_EXCEPTION_ARN,
      })
      .promise()

    console.info('OccurrenceExpiredHandler: sent message to sns', {
      snsResponse: JSON.stringify(snsResponse),
    })

    const {
      applicationId,
      scheduleId,
    } = updatedOccurrenceDocument as schemas.ScheduleOccurrenceDocumentSchema

    console.debug('OccurrenceExpiredHandler: invoking step function', {
      scheduleId,
      stateMachineArn: process.env.ENSURE_OCCURRENCES_STEP_FUNCTION,
    })

    await sfs
      .startExecution({
        input: JSON.stringify({ scheduleId }),
        name: `${applicationId}-${scheduleId}-${Date.now()}`,
        stateMachineArn: process.env.ENSURE_OCCURRENCES_STEP_FUNCTION,
      })
      .promise()
  } catch (err) {
    console.error('OccurrenceExpiredHandlerError', {
      err,
      event: JSON.stringify(event),
    })

    if (errors.isKnownError(err)) {
      throw err
    }

    throw new errors.UnknownError()
  }
}
