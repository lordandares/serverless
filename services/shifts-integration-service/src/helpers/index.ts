export { default as getPlatformEndpoint } from './get-platform-endpoint'
export { default as getSnsPayload } from './get-sns-payload'
export {
  getWinTeamMessages,
  getWinTeamTimePunch,
} from './get-winteam-time-punch'
export { default as notifyUser } from './notify-user'
export { default as parseWinTeamMessages } from './parse-winteam-messages'
export { default as sendMessage } from './send-message'
export {
  default as validateEndpointAttributes,
} from './validate-endpoint-attributes'
