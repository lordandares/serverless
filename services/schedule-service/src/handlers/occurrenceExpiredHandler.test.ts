import { errors } from '@lighthouse/serverless-common'
import * as AWS from 'aws-sdk-mock'
import * as MockDate from 'mockdate'
import { occurrenceExpiredHandler } from './occurrenceExpiredHandler'

import { createContext, createSnsEvent } from '../../../../__test__/helpers'
import { occurrenceDocument } from '../__fixtures__'
import { OCCURRENCE_STATUS_EXPIRED } from '../constants'

jest.mock('./lib/TimerService')

const timestamp = '2019-01-01T01:00:00.000Z'

beforeEach(() => {
  process.env.CREATE_EXCEPTION_ARN = 'create-exception'
  process.env.ENSURE_OCCURRENCES_STEP_FUNCTION =
    'ensure-occurrences-step-function'
  process.env.TABLE_SCHEDULES = 'table-schedules'

  MockDate.set(timestamp)
})

test('application error when body missing', async () => {
  expect.assertions(2)

  const updateSpy = jest
    .fn()
    .mockRejectedValue(new Error('ConditionalCheckFailedException'))

  AWS.mock('DynamoDB.DocumentClient', 'update', updateSpy)

  const body = null
  const context = createContext()

  try {
    await occurrenceExpiredHandler(createSnsEvent({ body }), context)
  } catch (err) {
    await expect(err).toBeInstanceOf(errors.ApplicationError)

    await expect(err).toHaveProperty(
      'message',
      'occurrenceExpiredHandler: missing body on event',
    )

    AWS.restore()
  }
})

test('application error when occurrence missing', async () => {
  expect.assertions(1)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }
  const context = createContext()

  const getSpy = jest.fn().mockResolvedValue({ Item: null })

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  try {
    await occurrenceExpiredHandler(createSnsEvent({ body }), context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[ResourceNotFoundError: The Resource "occurrence" with ID of "{"pk":"application1-occurrence","sk":"2019-01-01T00:00-occurrence1"}" could not be found]`,
    )
  }

  AWS.restore()
})

test('unknown error', async () => {
  expect.assertions(1)

  const getSpy = jest
    .fn()
    .mockResolvedValue({ Item: { ...occurrenceDocument, status: 'active' } })
  const updateSpy = jest.fn().mockRejectedValue(new Error('Unknown error'))

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'update', updateSpy)

  const pk = 'application1-schedule'
  const sk = '2019-01-01T00:00:00.000Z-occurrence1'
  const body = { pk, sk }
  const context = createContext()

  try {
    await occurrenceExpiredHandler(createSnsEvent({ body }), context)
  } catch (error) {
    expect(error).toMatchInlineSnapshot(
      `[UnknownError: Something went wrong! Try again or contact support if the problem persists.]`,
    )
  }

  AWS.restore()
})

test('skips occurrence if status is not active', async () => {
  expect.assertions(3)

  const body = {
    pk: 'application1-occurrence',
    sk: '2019-01-01T00:00-occurrence1',
  }
  const context = createContext()

  const getSpy = jest
    .fn()
    .mockResolvedValue({ Item: { ...occurrenceDocument, status: 'resolved' } })
  const startExecutionSpy = jest.fn().mockResolvedValue({})
  const updateSpy = jest
    .fn()
    .mockResolvedValue({ Attributes: occurrenceDocument })

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)

  await occurrenceExpiredHandler(createSnsEvent({ body }), context)

  expect(getSpy.mock.calls).toMatchSnapshot()
  expect(updateSpy).toBeCalledTimes(0)
  expect(startExecutionSpy).toBeCalledTimes(0)

  AWS.restore()
})

test('sets occurrence to expired and invokes ensure occurrences step function', async () => {
  expect.assertions(4)

  const getSpy = jest
    .fn()
    .mockResolvedValue({ Item: { ...occurrenceDocument, status: 'active' } })
  const snsSpy = jest.fn().mockResolvedValue({ MessageId: 'message-uid' })
  const startExecutionSpy = jest.fn().mockResolvedValue({})
  const updateSpy = jest
    .fn()
    .mockResolvedValue({ Attributes: occurrenceDocument })

  AWS.mock('DynamoDB.DocumentClient', 'get', getSpy)
  AWS.mock('DynamoDB.DocumentClient', 'update', updateSpy)
  AWS.mock('SNS', 'publish', snsSpy)
  AWS.mock('StepFunctions', 'startExecution', startExecutionSpy)

  const pk = 'application1-schedule'
  const sk = '2019-01-01T00:00:00.000Z-occurrence1'
  const body = { pk, sk }
  const context = createContext()

  const result = await occurrenceExpiredHandler(
    createSnsEvent({ body }),
    context,
  )

  expect(updateSpy).toHaveBeenCalledWith(
    {
      AttributeUpdates: {
        expiredAt: {
          Action: 'PUT',
          Value: timestamp,
        },
        expiredBy: {
          Action: 'PUT',
          Value: 'invokedFunctionArn',
        },
        status: {
          Action: 'PUT',
          Value: OCCURRENCE_STATUS_EXPIRED,
        },
        updatedAt: {
          Action: 'PUT',
          Value: timestamp,
        },
        updatedBy: {
          Action: 'PUT',
          Value: {
            id: 'awsRequestId',
            label: 'invokedFunctionArn',
            type: 'system',
          },
        },
      },
      Key: { pk, sk },
      ReturnValues: 'ALL_NEW',
      TableName: process.env.TABLE_SCHEDULES,
    },
    expect.any(Function),
  )

  expect(result).toMatchSnapshot()
  expect(snsSpy.mock.calls).toMatchSnapshot()
  expect(startExecutionSpy.mock.calls).toMatchSnapshot()

  AWS.restore()
})
