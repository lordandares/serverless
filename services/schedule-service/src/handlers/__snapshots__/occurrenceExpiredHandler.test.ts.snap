// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets occurrence to expired and invokes ensure occurrences step function 1`] = `undefined`;

exports[`sets occurrence to expired and invokes ensure occurrences step function 2`] = `
Array [
  Array [
    Object {
      "Message": "{\\"applicationId\\":\\"application1\\",\\"createdAt\\":\\"2019-01-01T00:00:00.000Z\\",\\"createdBy\\":{\\"id\\":\\"user1\\",\\"label\\":\\"Unknown User\\",\\"type\\":\\"user\\"},\\"data\\":{\\"occurrenceInterval\\":[1,2],\\"scheduleName\\":\\"Schedule 1\\",\\"serviceInterval\\":[1,2],\\"timezone\\":\\"Australia/Melbourne\\",\\"type\\":\\"visit\\"},\\"endAt\\":\\"2019-02-01T23:59:59.999Z\\",\\"groupType\\":\\"occurrence\\",\\"locationId\\":\\"5d163137b8b3b7000127edd1\\",\\"location_endAt_occurrenceId\\":\\"location1-2019-02-02T00:00:00.000Z-occurrence1\\",\\"occurrenceId\\":\\"occurrence1\\",\\"pk\\":\\"application1-occurrence\\",\\"scheduleId\\":\\"schedule1\\",\\"sk\\":\\"2019-01-01T00:00-occurrence1\\",\\"startAt\\":\\"2019-02-01T00:00:00.000Z\\",\\"status\\":\\"pending\\",\\"updatedBy\\":{\\"id\\":\\"user1\\",\\"label\\":\\"Unknown User\\",\\"type\\":\\"user\\"}}",
      "TopicArn": "create-exception",
    },
    [Function],
  ],
]
`;

exports[`sets occurrence to expired and invokes ensure occurrences step function 3`] = `
Array [
  Array [
    Object {
      "input": "{\\"scheduleId\\":\\"schedule1\\"}",
      "name": "application1-schedule1-1546304400000",
      "stateMachineArn": "ensure-occurrences-step-function",
    },
    [Function],
  ],
]
`;

exports[`skips occurrence if status is not active 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "pk": "application1-occurrence",
        "sk": "2019-01-01T00:00-occurrence1",
      },
      "TableName": "table-schedules",
    },
    [Function],
  ],
]
`;
